examples/cmd/logtransport.go

```diff
@@ -22,15 +21,23 @@ type logTransport struct {
 func (t *logTransport) RoundTrip(r *http.Request) (*http.Response, error) {
 	t0 := time.Now()
 
-	reqBody, _ := ioutil.ReadAll(r.Body)
-	r.Body = ioutil.NopCloser(bytes.NewBuffer(reqBody))
-
+	// Since the underlaying Transport.RoundTrip might consume the Request, and it's also needed later to log the process,
+	// it is needed to make an snapshot of the Request.Body in advance; otherwise, its will be no longer accessible.
+	bakupReqBody := utils.BackupRequestBody(r)
 	resp, err := t.T.RoundTrip(r)
+	r.Body = bakupReqBody
+
+	// Since the outter Transport.RoundTrip calling this one might also consume the Request, and it is also needed here to log
+	// the process, it will be done over a snapshot of the Request.Body; otherwise, its will be no longer accessible by outer Transports.
+	reqBody, _ := utils.ReadRequestAndRestore(r)
+
 	if err != nil {
 		return resp, err
 	}
 
-	respBody, err := ioutil.ReadAll(resp.Body)
+	// Since the Response has been read, it is needed to restore it
```

Comment:
Since this is used as the deepest transport, the response shouldn't have been read yet. Maybe to make things more clear and decouple responsibilities, we could have a `preserveBodyTransport` used as the deepest one that only cares about wrapping request and response with `ioutil.NopCloser`. Something like this:
```
type preserveBodyTransport struct {
	T http.RoundTripper
}

func (t *preserveBodyTransport) RoundTrip(r *http.Request) (*http.Response, error) {
	reqBody, err := t.makeBodyImmortal(r.Body)
	if err != nil {
		return nil, err
	}

	r.Body = reqBody

	resp, err := t.T.RoundTrip(r)
	defer resp.Body.Close()
	if err != nil {
		return nil, err
	}

	respBody, err := t.makeBodyImmortal(resp.Body)
	if err != nil {
		return nil, err
	}

	resp.Body = respBody

	return resp, err
}

func (t *preserveBodyTransport) makeBodyImmortal(body io.ReadCloser) (io.ReadCloser, error) {
	b, err := ioutil.ReadAll(body)
	if err != nil {
		return nil, err
	}

	return ioutil.NopCloser(bytes.NewBuffer(b)), nil
}
```

Reply:
The closing of the body of the response is due to this: https://github.com/src-d/metadata-retrieval/pull/57/files/65707c484f77b1f82d5d062c07eb738a55f626d8#r335415611.

Reply:
BTW I didn't tested it.

Reply:
Nope, sorry that wouldn't work as you still won't be able to read buffer twice, but at least it won't be closed.

Reply:
If I understood you, any body built with `ioutil.NopCloser` would let you read it at will, as many times you'd want.
But I could not reproduce it here: https://play.golang.org/p/GCrkec2Hx0w

```go
func main() {
	immortalBody := ioutil.NopCloser(bytes.NewBuffer([]byte("something")))
	
	// you consume the body
	content, err := ioutil.ReadAll(immortalBody)
	fmt.Println("1st. pass", string(content), err)
	
	// then you try reading it again...
	content, err = ioutil.ReadAll(immortalBody)
	fmt.Println("2nd. pass", string(content), err)
}
```

Reply:
From your latest comment, I assume we can collapse this discussion, and continue the one about closing the `Request.Body` at https://github.com/src-d/metadata-retrieval/pull/57#discussion_r335415611

Feel free to uncollapse :pray: if we should still discuss something more, or if I should consider any other suggestion about this topic.

Reply:
nope, in fact initially I thought that it would solve, but it just make it avoid the body to be closed, just ignore this.